packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.test.ts

```diff
@@ -0,0 +1,645 @@
+          const mappedAndSortedActions =
+            mapAndSortActionsByMostRecent(mockRawActions);
+          const ticketCreationDate = JSON.stringify(
+            createRelativeDate(caseCreationDate, 120, 0).getTime()
+          );
+
+          const sessions = splitActionsIntoTimelineSessions(
+            mappedAndSortedActions,
+            ticketCreationDate
+          );
+
+          expect(sessions.session?.actions.length).toEqual(1);
+          expect(sessions.session?.actions[0].actionType).toEqual(
+            UserActionType.TICKET_CREATION
+          );
+          expect(sessions.session?.actions[0].timestamp).toEqual(
+            ticketCreationDate
+          );
+          expect(sessions).toEqual(sessions);
+        });
+      });
+    });
+  });
+
+  describe('#isActionWithinSessionThreshold', () => {
+    it('should return true if the action is within the session inactivity threshold of the previous', async () => {
+      const action = {
+        actionType: 'CLICK' as UserActionType,
+        timestamp: firstSessionDate.getTime().toString(),
+        eventData: {},
+        searchHub: 'in-product-help',
+        document: {title: 'title', uriHash: 'TtnKwc0Lo2GY9WAi'},
+      };
+
+      // Added 1 min to action timestamp, so it is within 30 mins of the previous action
+      const previousEndDateTime = JSON.stringify(
+        createRelativeDate(firstSessionDate, 1, 0).getTime()
+      );
+
+      const isSameSession = isActionWithinSessionThreshold(
+        action,
+        previousEndDateTime
+      );
+      expect(isSameSession).toEqual(true);
+    });
+
+    it('should return false if the action is not within 30mins of the previous', async () => {
+      const action = {
+        actionType: 'CLICK' as UserActionType,
+        timestamp: firstSessionDate.getTime().toString(),
+        eventData: {},
+        searchHub: 'in-product-help',
+        document: {title: 'title', uriHash: 'TtnKwc0Lo2GY9WAi'},
+      };
+
+      // Added 1 hour to action timestamp, so it is not within 30 mins of the previous action
+      const previousEndDateTime = JSON.stringify(
+        createRelativeDate(firstSessionDate, 60, 0).getTime()
+      );
+
+      const isSameSession = isActionWithinSessionThreshold(
+        action,
+        previousEndDateTime
+      );
+      expect(isSameSession).toEqual(false);
+    });
+  });
+
+  describe('#filterTimelineActions', () => {
+    it('should properly filter out the timeline of the actions that are to be excluded and of empty searches', async () => {
+      const actionsToExclude = ['useless_event'];
+      const mockRawActions = [...fakeActions].slice(0, 8);
+      const mappedAndSortedActions =
+        mapAndSortActionsByMostRecent(mockRawActions);
+      const ticketCreationDate = JSON.stringify(caseCreationDate.getTime());
+      const sessionsTimeline = splitActionsIntoTimelineSessions(
+        mappedAndSortedActions,
+        ticketCreationDate
+      );
```

Comment:
We should not do these steps at the beginning of the tests, cause this is a unit test it should only depend  on `filterTimelineActions`  and not the other functions, this remark can be applied for all the other test as well.


This should halve the following structure:
```
// manually create a timeline
const timeline = { 
   followingSessions: [...],
   session: {...}
   precedingSessions: [...],
}
// call filterTimelineActions on  timeline
const filteredTimeline = filterTimelineActions(sessionsTimeline, actionsToExclude);

// compare result with expected timeline (you can use a snapshot to do this comparaison)
expect(filteredTimeline).toMatchSnapshot();
// OR
expect(filteredTimeline).toEqual(expectedFiltredTimeline);

```

Reply:
Thats a pattern etienne told me to do so since we know these methods are working and that they are tested at that point. This is only making sure even more that the previous methods work as expected. Saves us also from having to manually build static timelines, actions, etc. every time.

Reply:
the problem with the logic is that if the test fail u can't know for sure if the problem is on `mapAndSortActionsByMostRecent ` or `splitActionsIntoTimelineSessions ` or in the actual `filterTimelineActions` 

which the strength of a unit test as it allows to isolate the function we wanna test from anything that can impact it.

Reply:
I get that, but it also prevents us from hardcoding a lot of stuff which would crowd the file while testing even more some of the functions.

Reply:
Crowding the file is an important concern, but it is not necessary our main concern when it comes to test.
The main concern remains to create useful robust readable tests that test each function in isolation, that's the definition of a unit test.

when we start calling multiple functions in our test, it kinda becomes more of an integration test rather than a unit test.

Additionally, a way that would help us prevent crowding the file is having snapshots of the expected data in stead of hardcoding it in the file.